lib/SILOptimizer/Utils/CanonicalOSSALifetime.cpp

```diff
@@ -411,149 +424,162 @@ void CanonicalizeOSSALifetime::extendLivenessThroughOverlappingAccess() {
 // liveness computed in Step 1.
 //===----------------------------------------------------------------------===//
 
-// Look past ignoreable instructions to find a destroy on \p edgeBB that
-// destroys \p def.
-static DestroyValueInst *findDestroyOnCFGEdge(SILBasicBlock *edgeBB,
-                                              SILValue def) {
-  for (auto &inst : *edgeBB) {
-    if (CanonicalizeOSSALifetime::ignoredByDestroyHoisting(inst.getKind()))
-      continue;
-    if (auto *destroy = dyn_cast<DestroyValueInst>(&inst)) {
-      if (destroy->getOperand() == def)
-        return destroy;
-      continue;
+namespace swift {
+/// Extends the boundary from PrunedLiveness down to preexisting destroys of the
+/// def which aren't separated from the original boundary by "interesting"
+/// instructions.
+///
+/// The motivation for extending the boundary is to avoid "churning" when
+/// iterating to a fixed point by canonicalizing the lifetimes of several
+/// values with overlapping live ranges and failing to find a fixed point
+/// because their destroys are repeatedly hoisted over one another.
+class CanonicalizeOSSALifetimeBoundaryExtender {
+  SSAPrunedLiveness &liveness;
+  PrunedLivenessBoundary &originalBoundary;
+  SILValue currentDef;
+
+public:
+  CanonicalizeOSSALifetimeBoundaryExtender(
+      SSAPrunedLiveness &liveness, PrunedLivenessBoundary &originalBoundary,
+      SILValue currentDef)
+      : liveness(liveness), originalBoundary(originalBoundary),
+        currentDef(currentDef){};
+
+  /// Compute the extended boundary by walking out from the original boundary
+  /// (from PrunedLiveness::computeBoundary) down to any destroys that appear
+  /// later but which aren't separated from the original boundary by
+  /// "interesting" users.
+  void extend(CanonicalizeOSSALifetimeBoundary &boundary) {
+    for (auto *def : originalBoundary.deadDefs) {
+      SILInstruction *instruction = nullptr;
+      if (auto *argument = dyn_cast<SILArgument>(def)) {
+        instruction = &*argument->getParent()->begin();
```

Comment:
I can't follow the logic of substituting an argument with the first instruction in a block and treating that like a user. Seems like it will give you the wrong insertion point anyway. If you instead called `extendBoundaryFromCFGEdge` for dead arguments, I could understand that.

Reply:
Thank you!  Fixed and added a test.